src/transformers/parallelization_utils.py

```diff
@@ -0,0 +1,1098 @@
+
+        Args:
+            inputs (Tensor): input tensor
+
+        Returns:
+            Tensor: gathered tensor
+        """
+
+        if self._enable_grad(inputs):
+            return self._gather_fn.apply(inputs)
+        else:
+            return self._gather(inputs)
+
+    @staticmethod
+    def _enable_grad(inputs: Tensor) -> bool:
+        """
+        Check current tensor is enabled to pass gradient.
+
+        Args:
+            inputs (Tensor): input tensor
+
+        Returns:
+            bool: whether gradient can be passed or not
+        """
+        return torch.is_grad_enabled() and inputs.requires_grad
+
+    def _initialize_functions(self):
+        class Broadcast(Function):
+            @staticmethod
+            def forward(ctx, inputs):
+                return self._broadcast(inputs)
+
+            @staticmethod
+            def backward(ctx, inputs):
+                return self._reduce(inputs)
+
+        class Reduce(Function):
+            @staticmethod
+            def forward(ctx, inputs):
+                return self._reduce(inputs)
+
+            @staticmethod
+            def backward(ctx, inputs):
+                return self._broadcast(inputs)
+
+        class Scatter(Function):
+            @staticmethod
+            def forward(ctx, inputs):
+                return self._scatter(inputs)
+
+            @staticmethod
+            def backward(ctx, inputs):
+                return self._gather(inputs)
+
+        class Gather(Function):
+            @staticmethod
+            def forward(ctx, inputs):
+                return self._gather(inputs)
+
+            @staticmethod
+            def backward(ctx, inputs):
+                return self._scatter(inputs)
+
+        return {
+            "broadcast": Broadcast,
+            "reduce": Reduce,
+            "scatter": Scatter,
+            "gather": Gather,
+        }
+
+    @staticmethod
+    def initialize_distributed():
+        """Initialize torch.distributed and mpu."""
+        if not torch.distributed.is_initialized():
+            rank = int(os.getenv("RANK", 0))
+            world_size = int(os.getenv("WORLD_SIZE", 1))
+            device_count = torch.cuda.device_count()
+
+            if device_count > 0:
+                device = rank % device_count
+                torch.cuda.set_device(device)
+
+            init_method = "tcp://"
+            master_ip = os.getenv("MASTER_ADDR", "localhost")
+            master_port = os.getenv("MASTER_PORT", "6000")
+            init_method += master_ip + ":" + master_port
+            torch.distributed.init_process_group(
+                backend="nccl",
+                world_size=world_size,
+                rank=rank,
+                init_method=init_method,
+            )
+
+
+@dataclass
+class ParallelLayer:
+    name: str = None
+    weight: torch.Tensor = None
+    bias: torch.Tensor = None
+    replace: dict = None
+    n_fused: int = None
+    reversed: bool = None
+    scale_attention: bool = None
```

Comment:
It's for kernel fusion. I will add kernel fusion later.
@RezaYazdaniAminabadi @stas00